### PR TITLE
Emails

### PR DESCRIPTION
**Pull request by mgax**
Headers, attachments, șaorma cu de toate.

### CODE REVIEW

File: snoop/data/analyzers/email.py [hunk 1]
```diff
@@ -26,14 +27,56 @@ def get_headers(message):
...
+        payload_bytes = message.get_payload(decode=True)
+        with models.Blob.create() as writer:
+            writer.write(payload_bytes)
+        rmeta_blob = tika.rmeta(writer.blob)
+        with rmeta_blob.open(encoding='utf8') as f:
+            rmeta_data = json.load(f)
+        rv['text'] = rmeta_data[0]['X-TIKA:content']
+
+    if message.is_multipart():
+        rv['parts'] = [dump_part(part) for part in message.get_payload()]
```
Comment (gabriel-v) [inline snoop/data/analyzers/email.py]: If you recurse into `dump_part` for parts of the message, `dump_part` will try to extract the headers from those: see the first line from `dump_part`. I don't think it's a good idea to look for headers where there shouldn't be any.

File: snoop/data/analyzers/email.py [hunk 2]
```diff
@@ -26,14 +27,56 @@ def get_headers(message):
...
+    return rv
+
+
 @shaorma('email.parse')
 def parse(blob):
     with blob.open() as f:
-        message = email.message_from_bytes(f.read())
+        message_bytes = f.read()
+
+    if message_bytes[:3] == b'\xef\xbb\xbf':
```
Comment (gabriel-v) [inline snoop/data/analyzers/email.py]: What is this? Please include the words `byte order mark, thanks Microsoft` in a comment above this.

File: snoop/data/filesystem.py [hunk 3]
```diff
@@ -139,3 +146,47 @@ def create_file(parent_directory, name, original):
...
+        for attachment in attachments:
+            original = models.Blob.objects.get(pk=attachment['blob_pk'])
+            size = original.path().stat().st_size
+            now = timezone.now()
+
+            (file, _) = attachments_dir.child_file_set.get_or_create(
+                name=attachment['name'],
+                defaults=dict(
+                    collection=file.collection,
+                    ctime=now,
```
Comment (gabriel-v) [inline snoop/data/filesystem.py]: I think these (`ctime`, `mtime`) should be a time taken from the email, not the current digest time. Or, if we can figure out their actual create/modify times from the email itself? If we can't figure them out, then just inherit them from the email file itself (the first one? the last one?)

File: snoop/data/digests.py [hunk 4]
```diff
@@ -90,46 +90,88 @@ def parent_id(file):
...
+        if attachments:
+            rv['children'] = [
+                child_file(f)
+                for f in attachments.child_file_set.order_by('pk').all()
+            ]
+
+    return rv
+
+
+def child_file(file):
```
Comment (gabriel-v) [inline snoop/data/digests.py]: There's nothing specific to a `child` file here: maybe name it `file_to_dict` or something?

File: snoop/data/digests.py [hunk 5]
```diff
@@ -90,46 +90,88 @@ def parent_id(file):
...
-        }
-
-    def child_dir(directory):
-        return {
-            'id': directory_id(directory),
-            'content_type': 'application/x-directory',
-            'filename': directory.name,
-        }
 
+def child_dir(directory):
```
Comment (gabriel-v) [inline snoop/data/digests.py]: There's nothing specific to a `child` dir here: maybe name it `dir_to_dict` or something?

File: snoop/data/digests.py [hunk 6]
```diff
@@ -90,46 +90,88 @@ def parent_id(file):
...
+
+    text_bits = []
+    for part in iter_parts(email_data):
+        part_text = part.get('text')
+        if part_text:
+            text_bits.append(part_text)
+
+    headers_to = []
+    for header in ['To', 'Cc', 'Bcc', 'Resent-To', 'Recent-Cc', 'Reply-To']:
+        headers_to += headers.get(header, [])
```
Comment (gabriel-v) [inline snoop/data/digests.py]: maybe deduplicate these? And `Reply-to` isn't really a "to" field.

File: testsuite/test_integration.py [hunk 7]
```diff
@@ -58,5 +58,4 @@ def feed_page(url):
 
     # .partial.emlx
     partialemlx = docs[ID['partialemlx']]
-    headers = partialemlx['content']['_emailheaders']
-    assert headers['Subject'] == ["Re: promulgare lege"]
+    assert partialemlx['content']['subject'] == "Re: promulgare lege"
```
Comment (gabriel-v) [inline testsuite/test_integration.py]: could you also add a few of the above tested mails (maybe the most complex one) here in the integration test?

File: snoop/data/analyzers/email.py (hunk 1 above)
Comment (mgax) [reply snoop/data/analyzers/email.py]: We're going to need headers of sub-parts when we're going to decrypt pgp emails.

File: snoop/data/digests.py (hunk 5 above)
Comment (mgax) [reply snoop/data/digests.py]: It's specific to the "child" data structure which is part of the digest document:

```json
{
  "id": "foo",
  "content": {},
  "children": [{"id": "bar", "content_type": "text/plain", "filename": "bar.txt"}]
}
```

**Pull request merged**